Add ERC: ERC-20 Transfer Reference Extension

radeksvarz opened the pull request.

abcoathup commented on ERCS/erc-xyz.md:
```diff
@@ -0,0 +1,166 @@
+---
+eip: XXXX
```
```suggestion
eip: 7699
```
Assigning next sequential EIP/ERC/RIP number.
Numbering changed to sequential from 7500.

Please also update the filename.

abcoathup commented on ERCS/erc-xyz.md:
```diff
@@ -0,0 +1,166 @@
+title: ERC20 Payment reference extension
+description: A minimal standard interface for ERC20 tokens allowing users to include a unique identifier (payment reference) for each transaction to help distinguish and associate payments with orders and invoices.
+author: Radek Svarz (@radeksvarz)
+discussions-to: https://ethereum-magicians.org/t/erc-xxxx-erc20-payment-reference-extension/19826
```
```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7699-erc20-payment-reference-extension/19826
```

SamWilsn reviewed (commented):
You'll need to fix all the errors generated by the linter: https://github.com/ethereum/ERCs/actions/runs/8858664958/job/24327583845

radeksvarz commented:
> You'll need to fix all the errors generated by the linter: https://github.com/ethereum/ERCs/actions/runs/8858664958/job/24327583845

@SamWilsn Thx for letting me know. I will fix those. Would you be so kind and help initiate the discussion on related ethereum magicians among wallets (mainly Metamask, Rabby) and explorers (Etherscan, Blockscout)?

SamWilsn commented:
@radeksvarz we'd be happy to have you on the next AllWalletDevs call! You can find the details and discord server here: https://allwallet.dev/

SamWilsn commented on ERCS/erc-7699.md:
````diff
@@ -0,0 +1,162 @@
+## Reference Implementation
+
+```
+// SPDX-License-Identifier: MIT
````
I can't merge this until either: this license is changed to CC0-1.0, or the code is moved into the `../assets/erc-7699/` directory.

radeksvarz replied:
Thx, updated

radeksvarz commented:
I added the section ## Privacy Considerations (reflecting the Web3Privacy Now initiative). Is that a problem?

radeksvarz commented:
> I added the section ## Privacy Considerations (reflecting the Web3Privacy Now initiative). Is that a problem?

@SamWilsn I fixed the license, can you please react on the section issue?

radeksvarz commented:
@abcoathup , @fhar28 pls help

I added the section ## Privacy Considerations (reflecting the Web3Privacy Now initiative). Is that a problem?

SamWilsn commented on ERCS/erc-7699.md:
```diff
@@ -0,0 +1,212 @@
+
+The corresponding ERC-20 `Transfer` event MUST be emitted following the `TransferReference` event, ideally immediately afterwards for the client to be able to seek the associated `Transfer` event log record. 
+
+Emitted `loggedReference` MIGHT be the exact copy of the `transferReference` (when less then 33 bytes) or the derived data from the rich `transferReference` structure and other processing. This is up to the implementer. One MUST NOT expect the `transferReference` and `loggedReference` data are equal.
```
Is "MIGHT" one of the keywords? I don't actually know.

radeksvarz replied:
@SamWilsn The keyword is "MAY" . Should I create a separate PR for these fixes , or edit this one?

radeksvarz replied:
fixed with #716  Thx

SamWilsn commented on ERCS/erc-7699.md:
```diff
@@ -0,0 +1,212 @@
+
+The corresponding ERC-20 `Transfer` event MUST be emitted following the `TransferReference` event, ideally immediately afterwards for the client to be able to seek the associated `Transfer` event log record. 
+
+Emitted `loggedReference` MIGHT be the exact copy of the `transferReference` (when less then 33 bytes) or the derived data from the rich `transferReference` structure and other processing. This is up to the implementer. One MUST NOT expect the `transferReference` and `loggedReference` data are equal.
```
```suggestion
Emitted `loggedReference` MIGHT be the exact copy of the `transferReference` (when less then 33 bytes) or the derived data from the rich `transferReference` structure and other processing. This is up to the implementer. One MUST NOT expect the `transferReference` and `loggedReference` data to be equal.
```

radeksvarz replied:
fixed with #716  Thx

SamWilsn commented on ERCS/erc-7699.md:
```diff
@@ -0,0 +1,212 @@
+
+The `transferReference` parameter is not limited in length by design, users are motivated to keep it short due to calldata and execution gas costs.
+
+The `TransferReference` event MUST NOT be anonymous. This is to ensure that the event signature is logged and can be used as a filter.
```
This _might_ be clear to someone reading this slower than I am, but I feel "anonymous" here is a bit vague. Might be something to fix while in Draft.

radeksvarz replied:
ah, it refers to the type of event, resp. to the specifier as documented here: https://docs.soliditylang.org/en/latest/contracts.html#events

Would this be better?
"""
The `TransferReference` event MUST NOT be declared with the `anonymous` specifier.
"""

radeksvarz replied:
fixed with #716  Thx

SamWilsn commented on ERCS/erc-7699.md:
```diff
@@ -0,0 +1,212 @@
+
+### Privacy Considerations
+
+Reference data privacy: Including payment references in token transfers may expose sensitive information about the transaction or the parties involved. Implementers and users SHOULD carefully consider the privacy implications and ensure that payment references do not reveal sensitive information. To mitigate this risk, implementers can consider using encryption or other privacy-enhancing techniqu ...
```
```suggestion
Reference data privacy: Including payment references in token transfers may expose sensitive information about the transaction or the parties involved. Implementers and users should carefully consider the privacy implications and ensure that payment references do not reveal sensitive information. To mitigate this risk, implementers can consider using encryption or other privacy-enhancing techniques to protect payment reference data.
```

Avoid using the keywords outside the specification section.

radeksvarz replied:
fixed with #716  Thx

SamWilsn reviewed (approved):
Feel free to fix these at some point while you're in draft.
